File: Sources/NockerlDesign/Components/NockerlLogo.swift
// NockerlLogo: the canonical three-peaks Nockerl mark, EXACTLY as it ships in the
// apps (Voice NockerlLogo.swift, Android ic_nockerl_logo*.xml, web NockerlLogo.tsx):
// three OVERLAPPING triangles in three DISTINCT grayscale shades (the shade
// difference gives the layered, dimensional look; a flat fill is WRONG). Identical
// geometry on the `16 20 64 56` viewBox, a y=72 baseline, drawn left → center →
// right. MONOCHROME grayscale, recolored per theme, NEVER cyan (law §11). Mirrors
// the Compose NockerlLogo.

import SwiftUI

/// The forced ink ladder, named by the surface the mark sits ON.
public enum NockerlLogoTone: Equatable {
    /// For a DARK surface: LIGHT ink.
    case onDark
    /// For a LIGHT surface: DARK ink.
    case onLight

    /// The three peak shades (left, center, right) for this ladder.
    var ink: (Color, Color, Color) {
        switch self {
        case .onDark: return (Color(white: 0.898), Color(white: 1.0), Color(white: 0.8))
        case .onLight: return (Color(white: 0.333), Color(white: 0.133), Color(white: 0.467))
        }
    }
}

/// The Nockerl mark (see the file header). Theme-aware by default; pass a [tone] to
/// force a ladder.
public struct NockerlLogo: View {
    private let size: CGFloat
    private let tone: NockerlLogoTone?
    private let accessibilityLabel: String?

    /// Native mark width (viewBox `16 20 64 56` → 64×56, an 8:7 ratio).
    static let nativeW: CGFloat = 64
    /// Native mark height.
    static let nativeH: CGFloat = 56

    /// Create the mark.
    /// - Parameters:
    ///   - size: the mark HEIGHT; width keeps the native 8:7 ratio.
    ///   - tone: force an ink ladder, or `nil` (default) for theme-aware.
    ///   - accessibilityLabel: the a11y name, or `nil` when a sibling wordmark names
    ///     it (then the mark is decorative).
    public init(
        size: CGFloat = 24,
        tone: NockerlLogoTone? = nil,
        accessibilityLabel: String? = "Nockerl"
    ) {
        self.size = size
        self.tone = tone
        self.accessibilityLabel = accessibilityLabel
    }

    public var body: some View {
        LogoBody(size: size, tone: tone, accessibilityLabel: accessibilityLabel)
    }

    private struct LogoBody: View {
        let size: CGFloat
        let tone: NockerlLogoTone?
        let accessibilityLabel: String?

        @Environment(\.colorScheme) private var colorScheme

        var body: some View {
            let resolved = tone ?? (colorScheme == .dark ? .onDark : .onLight)
            let ink = resolved.ink
            let width = size * (NockerlLogo.nativeW / NockerlLogo.nativeH)

            Canvas { context, canvasSize in
                let sx = canvasSize.width / NockerlLogo.nativeW
                let sy = canvasSize.height / NockerlLogo.nativeH
                func pt(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
                    CGPoint(x: (x - 16) * sx, y: (y - 20) * sy)
                }
                func tri(_ a: CGPoint, _ b: CGPoint, _ c: CGPoint) -> Path {
                    var p = Path()
                    p.move(to: a)
                    p.addLine(to: b)
                    p.addLine(to: c)
                    p.closeSubpath()
                    return p
                }
                // left → center → right (paint order = depth order).
                context.fill(tri(pt(20, 72), pt(36, 36), pt(52, 72)), with: .color(ink.0))
                context.fill(tri(pt(32, 72), pt(48, 24), pt(64, 72)), with: .color(ink.1))
                context.fill(tri(pt(44, 72), pt(60, 32), pt(76, 72)), with: .color(ink.2))
            }
            .frame(width: width, height: size)
            .accessibilityHidden(accessibilityLabel == nil)
            .accessibilityLabel(accessibilityLabel ?? "")
            .accessibilityAddTraits(.isImage)
        }
    }
}

#if os(macOS)
import AppKit

public extension NockerlLogo {
    /// The mark as a monochrome **template** `NSImage` for AppKit status-item
    /// hosts (`MenuBarExtra` labels, `NSStatusItem` buttons), the one surface a
    /// SwiftUI view cannot reach. Apple's HIG: menu-bar icons must be
    /// template images so the system tints them for light/dark, vibrancy, and
    /// the selection highlight.
    ///
    /// Template rendering keeps only the ALPHA channel, so the mark's grayscale
    /// depth ladder (the layered look; a flat fill is WRONG, see the header)
    /// survives as per-peak alpha: left `0.898` · center `1.0` · right `0.8`,
    /// the `onDark` ink ladder expressed in the template's one monochrome
    /// channel. Peaks paint in the canonical left → center → right order with
    /// paint-order (replace, never blend) overlap semantics, exactly like the
    /// in-app mark.
    ///
    /// - Parameter size: the mark HEIGHT in points (width keeps the native 8:7
    ///   ratio). The default 16 sits the mark in the menu bar's ~18pt content
    ///   box, the status-item norm.
    /// - Returns: a template `NSImage` (`isTemplate = true`) the system tints.
    static func statusItemImage(size: CGFloat = 16) -> NSImage {
        let width = size * (NockerlLogo.nativeW / NockerlLogo.nativeH)
        // The three peaks on the canonical `16 20 64 56` viewBox (x1 y1 x2 y2
        // x3 y3), each with its depth-ladder shade: the onDark grayscale
        // ladder, carried as template alpha.
        let peaks: [(CGFloat, CGFloat, CGFloat, CGFloat, CGFloat, CGFloat, CGFloat)] = [
            (20, 72, 36, 36, 52, 72, 0.898),
            (32, 72, 48, 24, 64, 72, 1.0),
            (44, 72, 60, 32, 76, 72, 0.8),
        ]
        // flipped: true = a y-DOWN handler space, so the viewBox coordinates map
        // straight through (no manual flip).
        let image = NSImage(size: NSSize(width: width, height: size), flipped: true) { rect in
            guard let ctx = NSGraphicsContext.current else { return false }
            let sx = rect.width / NockerlLogo.nativeW
            let sy = rect.height / NockerlLogo.nativeH
            for (ax, ay, bx, by, cx, cy, shade) in peaks {
                let path = NSBezierPath()
                path.move(to: CGPoint(x: (ax - 16) * sx, y: (ay - 20) * sy))
                path.line(to: CGPoint(x: (bx - 16) * sx, y: (by - 20) * sy))
                path.line(to: CGPoint(x: (cx - 16) * sx, y: (cy - 20) * sy))
                path.close()
                // Opaque fill first: paint-order overlap with clean AA edges…
                ctx.compositingOperation = .sourceOver
                NSColor.black.setFill()
                path.fill()
                // …then thin the peak DOWN to its ladder alpha (destination-out
                // reduces uniformly and its edge rides the same AA boundary, so
                // overlaps never seam).
                if shade < 1 {
                    ctx.compositingOperation = .destinationOut
                    NSColor(white: 0, alpha: 1 - shade).setFill()
                    path.fill()
                }
            }
            ctx.compositingOperation = .sourceOver
            return true
        }
        image.isTemplate = true
        image.accessibilityDescription = "Nockerl"
        return image
    }
}
#endif

File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlLogo.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors

/**
 * The **Nockerl mark**: the canonical three-peaks silhouette, EXACTLY as it ships
 * in the apps (Voice `NockerlLogo.swift`, Android `ic_nockerl_logo*.xml`, web
 * `NockerlLogo.tsx`): three OVERLAPPING triangles in three DISTINCT grayscale
 * shades (the shade difference gives the mark its layered, dimensional look; a
 * single flat fill is WRONG). Identical geometry on the `16 20 64 56` viewBox, a
 * `y = 72` baseline, drawn left → center → right (left behind, right in front).
 *
 * MONOCHROME GRAYSCALE, recolored per theme, NEVER cyan (law §11: cyan lives only
 * in the lockup's product word):
 * - on a DARK surface → LIGHT ink (#E5E5E5 / #FFFFFF / #CCCCCC)
 * - on a LIGHT surface → DARK ink (#555555 / #222222 / #777777)
 *
 * Default is theme-aware (inferred from the resolved `canvas` luminance); pass a
 * [tone] to force a ladder (e.g. a mark pinned to a known-dark tile).
 *
 * @param modifier outer modifier.
 * @param size the mark HEIGHT; width keeps the native 8:7 ratio.
 * @param tone force an ink ladder, or `null` (default) for theme-aware.
 * @param contentDescription the a11y name, or `null` when a sibling wordmark
 *   already names it (then the mark is decorative).
 */
@Composable
fun NockerlLogo(
    modifier: Modifier = Modifier,
    size: Dp = 24.dp,
    tone: NockerlLogoTone? = null,
    contentDescription: String? = "Nockerl",
) {
    val colors = LocalNockerlColors.current
    val resolved =
        tone ?: if (colors.canvas.luminance() < 0.5f) NockerlLogoTone.ON_DARK else NockerlLogoTone.ON_LIGHT
    val ink = resolved.ink()
    val width = size * (MARK_NATIVE_W / MARK_NATIVE_H)

    val name = contentDescription
    val semantics =
        if (name != null) {
            Modifier.semantics {
                this.contentDescription = name
                this.role = Role.Image
            }
        } else {
            Modifier
        }

    Canvas(modifier = modifier.then(semantics).size(width = width, height = size)) {
        val sx = this.size.width / MARK_NATIVE_W
        val sy = this.size.height / MARK_NATIVE_H

        // Map a viewBox point (origin 16,20) into canvas space.
        fun pt(
            x: Float,
            y: Float,
        ) = Offset((x - 16f) * sx, (y - 20f) * sy)

        fun tri(
            a: Offset,
            b: Offset,
            c: Offset,
        ) = Path().apply {
            moveTo(a.x, a.y)
            lineTo(b.x, b.y)
            lineTo(c.x, c.y)
            close()
        }
        // left → center → right (paint order = depth order).
        drawPath(tri(pt(20f, 72f), pt(36f, 36f), pt(52f, 72f)), ink.first)
        drawPath(tri(pt(32f, 72f), pt(48f, 24f), pt(64f, 72f)), ink.second)
        drawPath(tri(pt(44f, 72f), pt(60f, 32f), pt(76f, 72f)), ink.third)
    }
}

/** The forced ink ladder, named by the surface the mark sits ON. */
enum class NockerlLogoTone {
    /** For a DARK surface: LIGHT ink. */
    ON_DARK,

    /** For a LIGHT surface: DARK ink. */
    ON_LIGHT,
    ;

    /** The three peak shades (left, center, right) for this ladder. */
    fun ink(): Triple<Color, Color, Color> =
        when (this) {
            ON_DARK -> Triple(Color(0xFFE5E5E5), Color(0xFFFFFFFF), Color(0xFFCCCCCC))
            ON_LIGHT -> Triple(Color(0xFF555555), Color(0xFF222222), Color(0xFF777777))
        }
}

/** Native mark width (viewBox is `16 20 64 56` → 64×56, an 8:7 ratio). */
private const val MARK_NATIVE_W = 64f

/** Native mark height. */
private const val MARK_NATIVE_H = 56f
